### ISSUE TITLE
Authentication Systems

### ISSUE THREAD

**Issue by the-maldridge** [BESTRobotics/registry#4]
My opinion is that we should use JWT's which can be used as bearer tokens as well as queried for what capabilities a user may have.  They will have a lifetime, but we can set this at something like an hour to make the renewal problem moot.

I don't think we should start with social login (oauth/gauth) as an option as that dramatically increases complexity, and until we can semi-permanently nail down the server name this operates under, it makes it very difficult to run that.

Its also worth pointing out that we're kind of inverting some flows that many BEST users are currently familiar with.  Most notably we're transitioning here from logging in "as the hub" to logging in as a user who may manage a hub.

**Comment by scauthen** [comment]
Does this mean when a hub director creates their account in this system, someone on the admin side will have to recognize them as a hub director assign them to a hub? This would mean that directors (or anyone currently using the dashboard) would have one login (the Google authentication) to access the Dashboard and a second login to access the Registry system. Having two different accounts may be challenging for some people.

**Comment by the-maldridge** [comment]
Our existing google authentication really isn't serviceable, and I'll be surprised if it works through the end of the year.  We can eventually support login via google accounts, but all that is is recognizing a secondary login flow (oauth2) and can be done after the fact.  I think its a reasonable request to have hub directors hold a second account for a period of time while we get the system setup.

You are correct in that for existing hub directors we'll need to backfill the data, then manually set them as administrators of their hubs.

@gdyoung93 can explain more, but I believe the long term goal is to fold most of the functionality of the dashboard into a replacement system that is more maintainable.

**Comment by gdyoung93** [comment]
Yes, that is long-term plan. I'm not sure what you are referring to about
Google Auth going away?  I think it is already oauth2. But, regardless, it
will have to be updated if it changes. I haven't gotten any notice that it
will be changing.

2nd login is OK temporarily but not long-term.

Greg

**Comment by the-maldridge** [comment]
I'm referring to the kind of embedded click-bounce-auth button currently in use which already has some trouble validating the session.  Bouncing through that many domains is not something that oauth was designed to do.  I'm also fairly sure that we're on oauth, not oauth2.

I foresee that we'd need to sustain the 2nd login for ~6 months in the worst case.

**Comment by gdyoung93** [comment]
Yeah, that's a Google thing. It used to be much simpler. This latest
incarnation does have trouble and I'd love to fix it.

greg

**Comment by the-maldridge** [comment]
I'm looking into this for user authentication and registration.  It seems to do everything under the sun and so would work for our purposes, but the maintenance history isn't exactly what I'd like.  I think we may start out with users that have no authentication, then go from there so as to not stall out the project.

**Comment by the-maldridge** [comment]
Username/password working, goth integration remains to be done.
